### PR TITLE
network: about route message counter

### LINKED ISSUE #17396: Assertion 'link->route_messages > 0' failed at src/network/networkd-route.c:1140, function int route_handler(sd_netlink *, sd_netlink_message *, Link *)(). Aborting.
**systemd version the issue has been seen with**
 > latest master

**Used distribution**
 > Arch Linux

**Linux kernel version used** (`uname -a`)
 > 5.7.12

**CPU architecture issue was seen on**
 > x86_64

The `systemd-networkd-tests.py` testsuite occasionally fails due to failed assertion during `test_route_static`:

```
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Remembering route: dst: 2001:1234:5:8f63::1/128, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: local, proto: kernel, type: local
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Remembering updated address: 149.10.124.58/28 (valid forever)
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: link_check_ready(): static addresses are not configured.
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Remembering route: dst: 149.10.124.58/32, src: n/a, gw: n/a, prefsrc: 149.10.124.58, scope: host, table: local, proto: kernel, type: local
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Remembering route: dst: 149.10.124.63/32, src: n/a, gw: n/a, prefsrc: 149.10.124.58, scope: link, table: local, proto: kernel, type: broadcast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Remembering route: dst: 149.10.124.48/28, src: n/a, gw: n/a, prefsrc: 149.10.124.58, scope: link, table: main, proto: kernel, type: unicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Remembering route: dst: 149.10.124.48/32, src: n/a, gw: n/a, prefsrc: 149.10.124.58, scope: link, table: local, proto: kernel, type: broadcast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Addresses set
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Configuring route: dst: 202.54.1.3/32, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: main, proto: static, type: unreachable
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Configuring route: dst: 149.10.123.3/32, src: n/a, gw: n/a, prefsrc: n/a, scope: link, table: local, proto: static, type: broadcast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Configuring route: dst: 149.10.123.4/32, src: n/a, gw: n/a, prefsrc: n/a, scope: link, table: main, proto: static, type: multicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Configuring route: dst: 202.54.1.2/32, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: main, proto: static, type: blackhole
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Configuring route: dst: 149.10.123.2/32, src: n/a, gw: n/a, prefsrc: n/a, scope: link, table: local, proto: static, type: anycast
Oct 19 21:46:54 arch.localdomain systemd-resolved[160261]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/resolve1 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=3848 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Configuring route: dst: 2001:1234:5:9fff:ff:ff:ff:ff/128, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: main, proto: static, type: unicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Configuring route: dst: 149.10.123.1/32, src: n/a, gw: n/a, prefsrc: n/a, scope: host, table: local, proto: static, type: local
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Configuring route: dst: 192.168.1.2/32, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: main, proto: static, type: unicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Configuring route: dst: n/a, src: n/a, gw: n/a, prefsrc: n/a, scope: link, table: main, proto: static, type: unicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Configuring route: dst: 202.54.1.4/32, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: main, proto: static, type: prohibit
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Configuring route: dst: 192.168.1.1/32, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: main, proto: static, type: unicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Configuring route: dst: 169.254.0.0/16, src: n/a, gw: n/a, prefsrc: n/a, scope: link, table: main, proto: static, type: unicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Configuring route: dst: 149.10.124.64/32, src: n/a, gw: n/a, prefsrc: n/a, scope: link, table: main, proto: static, type: unicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Configuring route: dst: 2001:1234:5:8fff:ff:ff:ff:ff/128, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: main, proto: static, type: unicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Configuring route: dst: n/a, src: n/a, gw: 149.10.125.65, prefsrc: n/a, scope: global, table: main, proto: static, type: unicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Configuring route: dst: n/a, src: n/a, gw: 149.10.124.64, prefsrc: n/a, scope: global, table: main, proto: static, type: unicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Configuring route: dst: 192.168.10.1/32, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: main, proto: static, type: unicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Configuring route: dst: 2001:1234:5:7fff:ff:ff:ff:ff/128, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: main, proto: static, type: unicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Configuring route: dst: n/a, src: n/a, gw: 2001:1234:5:8fff:ff:ff:ff:ff, prefsrc: n/a, scope: global, table: main, proto: static, type: unicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Configuring route: dst: 192.168.10.2/32, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: main, proto: static, type: unicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Could not add route: File exists
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Could not set routes: File exists
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Failed
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: State changed: configuring -> failed
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1/link/_3361 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=65 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: Received remembered route: dst: 202.54.1.3/32, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: main, proto: static, type: unreachable
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Received remembered route: dst: 149.10.123.3/32, src: n/a, gw: n/a, prefsrc: n/a, scope: link, table: local, proto: static, type: broadcast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Received remembered route: dst: 149.10.123.4/32, src: n/a, gw: n/a, prefsrc: n/a, scope: link, table: main, proto: static, type: multicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: Received remembered route: dst: 202.54.1.2/32, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: main, proto: static, type: blackhole
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Received remembered route: dst: 149.10.123.2/32, src: n/a, gw: n/a, prefsrc: n/a, scope: link, table: local, proto: static, type: anycast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Received remembered route: dst: 2001:1234:5:9fff:ff:ff:ff:ff/128, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: main, proto: static, type: unicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Received remembered route: dst: 149.10.123.1/32, src: n/a, gw: n/a, prefsrc: n/a, scope: host, table: local, proto: static, type: local
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Received remembered route: dst: 192.168.1.2/32, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: main, proto: static, type: unicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Received remembered route: dst: n/a, src: n/a, gw: n/a, prefsrc: n/a, scope: link, table: main, proto: static, type: unicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: Received remembered route: dst: 202.54.1.4/32, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: main, proto: static, type: prohibit
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Received remembered route: dst: 192.168.1.1/32, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: main, proto: static, type: unicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Received remembered route: dst: 169.254.0.0/16, src: n/a, gw: n/a, prefsrc: n/a, scope: link, table: main, proto: static, type: unicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Received remembered route: dst: 149.10.124.64/32, src: n/a, gw: n/a, prefsrc: n/a, scope: link, table: main, proto: static, type: unicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Received remembered route: dst: 2001:1234:5:8fff:ff:ff:ff:ff/128, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: main, proto: static, type: unicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Received remembered route: dst: n/a, src: n/a, gw: 149.10.125.65, prefsrc: n/a, scope: global, table: main, proto: static, type: unicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Received remembered route: dst: n/a, src: n/a, gw: 149.10.124.64, prefsrc: n/a, scope: global, table: main, proto: static, type: unicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Received remembered route: dst: 192.168.10.1/32, src: n/a, gw: 149.10.124.59, prefsrc: n/a, scope: global, table: main, proto: static, type: unicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Received remembered route: dst: 192.168.10.1/32, src: n/a, gw: 149.10.124.60, prefsrc: n/a, scope: global, table: main, proto: static, type: unicast
Oct 19 21:46:54 arch.localdomain systemd-resolved[160261]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/resolve1 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=3849 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Received remembered route: dst: 2001:1234:5:7fff:ff:ff:ff:ff/128, src: n/a, gw: 2001:1234:5:8fff:ff:ff:ff:ff, prefsrc: n/a, scope: global, table: main, proto: static, type: unicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Received remembered route: dst: 2001:1234:5:7fff:ff:ff:ff:ff/128, src: n/a, gw: 2001:1234:5:9fff:ff:ff:ff:ff, prefsrc: n/a, scope: global, table: main, proto: static, type: unicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Received remembered route: dst: n/a, src: n/a, gw: 2001:1234:5:8fff:ff:ff:ff:ff, prefsrc: n/a, scope: global, table: main, proto: static, type: unicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Received remembered route: dst: 192.168.10.2/32, src: n/a, gw: 2001:1234:5:8fff:ff:ff:ff:ff, prefsrc: n/a, scope: global, table: main, proto: static, type: unicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: dummy98: Received remembered route: dst: 192.168.10.2/32, src: n/a, gw: 2001:1234:5:9fff:ff:ff:ff:ff, prefsrc: n/a, scope: global, table: main, proto: static, type: unicast
Oct 19 21:46:54 arch.localdomain systemd-networkd[168483]: Assertion 'link->route_messages > 0' failed at src/network/networkd-route.c:1140, function int route_handler(sd_netlink *, sd_netlink_message *, Link *)(). Aborting.
```

Most recently seen in #17395 and #17392.

Related job logs:
 * https://jenkins-systemd.apps.ocp.ci.centos.org/job/upstream-vagrant-archlinux/329/artifact//systemd-centos-ci/index.html
 * https://jenkins-systemd.apps.ocp.ci.centos.org/job/upstream-vagrant-archlinux-sanitizers/337/artifact//systemd-centos-ci/index.html

/cc @yuwata

### PR DESCRIPTION
**Pull request by yuwata**
Hopefully fixes #17396.

### CODE REVIEW

File: src/network/networkd-route.c [hunk 1]
```diff
@@ -299,9 +299,10 @@ void route_hash_func(const Route *route, struct siphash *state) {
                 siphash24_compress(&route->src, FAMILY_ADDRESS_SIZE(route->family), state);
 
                 siphash24_compress(&route->gw_family, sizeof(route->gw_family), state);
-                if (IN_SET(route->gw_family, AF_INET, AF_INET6))
+                if (IN_SET(route->gw_family, AF_INET, AF_INET6)) {
                         siphash24_compress(&route->gw, FAMILY_ADDRESS_SIZE(route->gw_family), state);
-
+                        siphash24_compress(&route->gw_weight, sizeof(route->gw_weight), state);
+                }
```
Comment (keszybz) [inline src/network/networkd-route.c:305]: Hmm, why {} ?

**Review by keszybz — COMMENTED** [review]
LGTM.

**Comment by keszybz** [comment]
bionic-i386 timed out in "upstream", semaphoreci failed to download keys.

File: src/network/networkd-route.c (hunk 1 above)
Comment (poettering) [reply src/network/networkd-route.c:305]: because there are now two expressios in the {}

**Pull request merged**